### PR TITLE
Improve formatting of module arguments

### LINKED ISSUE #1678: Bug: misaligned higher-order functor parameters
The following example formats to itself with `ocamlformat.0.18.0`:

```ocaml
module F1 (G : functor (_ : T) ->
  T) (A : sig
  val x : int
end) =
struct end

module F2 (G : functor (_ : T) ->
  T_________________________________________________________________________)
                                                                            (A : sig
  val x : int
end) =
struct end

module F3
    (G : functor
      (_ : T____________________________________________)
      (_ : T____________________________________________)
      ->
  T) (A : sig
  val x : int
end) =
struct end
```

It appears that using immediate `functor ..` types confuses OCamlformat's normal placement strategy for functor parameters, resulting in a variety of unusual indentation. I'm not sure entirely what I'd expect this to format to, but roughly speaking:
 - In the first case, `F1`, I'd expect `G` to not break, and `A` to either start on the subsequent line or dock after `G`.
 - For `F2`, I'd expect `A` to start much further to the left.
 - For `F3`, I'd expect the `T` result type of `G` to be indented by another couple of levels (and perhaps `A` to start on its own line, left-aligned with `G`).

I don't know whether these are known limitations of the placement strategy, in which case feel free to reclassify this issue as a feature request or close it altogether.

### PR DESCRIPTION
**Pull request by Julow**
This fixes the formatting of module arguments, for example:

```diff
@@ -4501,11 +4501,12 @@ let f (module M : S with type t = int) = {M.a= 0}
 
 let flag = ref false
 
-module F (S : sig
-  module type T
-end)
-(A : S.T)
-(B : S.T) =
+module F
+    (S : sig
+      module type T
+    end)
+    (A : S.T)
+    (B : S.T) =
 struct
   module X = (val if !flag then (module A) else (module B) : S.T)
 end
@@ -6914,9 +6915,11 @@ end
 
 type 'a tree = E | N of 'a tree * 'a * 'a tree
 
-module Bootstrap2 (MakeDiet : functor (X : ORD) ->
-  SET with type t = X.t tree and type elt = X.t) : SET with type elt = int =
-struct
+module Bootstrap2
+    (MakeDiet : functor
+      (X : ORD)
+      -> SET with type t = X.t tree and type elt = X.t) :
+  SET with type elt = int = struct
   type elt = int
 
   module rec Elt : sig
```

The code is rewritten to avoid unsafe box operations. The second commit implements ocp-indent compatibility, which changes the docking behavior and indentation.

edit: fix #1678

### CODE REVIEW

**Comment by gpetiot** [comment]
Apparently this fixes #1678

File: lib/Params.mli [hunk 1]
```diff
@@ -36,6 +37,15 @@ module Exp : sig
     -> Fmt.t
 end
 
+module Mod : sig
+  type args =
+    { dock: bool  (** Whether each argument's [pro] should be docked. *)
+    ; arg_psp: Fmt.t  (** Break before every arguments. *)
```
Comment (gpetiot) [inline lib/Params.mli:43]: What do you think of using a `Fmt.sp` here?

**Review by gpetiot — APPROVED** [review]
The new output looks great!

LGTM, thanks!

File: lib/Params.mli (hunk 1 above)
Comment (Julow) [reply lib/Params.mli:43]: That was my initial idea but it's `str " "` in some cases. I come to the conclusion that `Fmt.t` is generally better than `Fmt.sp`.

Comment (gpetiot) [reply lib/Params.mli:43]: You can use `Fmt.Blank` for `" "`

**Pull request merged**